File: 03-GildedRose/java/group1part2/src/main/java/com/capgemini/coedevon/codingdojo/gildedrose/GildedRose.java
package com.capgemini.coedevon.codingdojo.gildedrose;

import java.util.List;

@SuppressWarnings("javadoc")
public class GildedRose {

  private static final int DEADLINE_SELL_IN = 0;

  private static final int SELL_IN_UNIT = 1;

  private static final int PASSES_THRESHOLD_TRIPLE_QUALITY = 5;

  private static final int PASSES_THRESHOLD_DOUBLE_QUALITY = 10;

  private static final int MAX_QUALITY = 50;

  private static final int MIN_QUALITY = 0;

  private static final int QUALITY_UNIT = 1;

  private static final int DOUBLE_QUALITY_UNIT = 2;

  private static final String SULFURAS = "Sulfuras, Hand of Ragnaros";

  private static final String BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert";

  private static final String AGED_BRIE = "Aged Brie";

  private static final String CONJURED_ITEM = "Conjured Item";

  private List<Item> items;

  public GildedRose(List<Item> items) {
    this.items = items;
  }

  public void updateQuality() {

    for (Item item : this.items) {

      applyQualityRules(item);
      decreaseSellIn(item);
      applyExpiredRules(item);
    }
  }

  /**
   * @param item
   */
  private void applyExpiredRules(Item item) {

    if (hasExpired(item)) {
      if (isPasses(item)) {
        setZeroQuality(item);
      } else if (isAgedBrie(item)) {
        increaseQuality(item);
      } else if (isConjured(item)) {
        decreaseQuality(item, DOUBLE_QUALITY_UNIT);
      } else {
        decreaseQuality(item);
      }
    }
  }

  /**
   * @param item
   */
  private void applyQualityRules(Item item) {

    boolean belowThresholdDoubleQuality = item.getSellIn() <= PASSES_THRESHOLD_DOUBLE_QUALITY;
    boolean belowThresholdTripleQuality = item.getSellIn() <= PASSES_THRESHOLD_TRIPLE_QUALITY;

    if (isPasses(item)) {
      increaseQuality(item);

      if (belowThresholdDoubleQuality) {
        increaseQuality(item);
      }

      if (belowThresholdTripleQuality) {
        increaseQuality(item);
      }
    } else if (isAgedBrie(item)) {
      increaseQuality(item);
    } else if (isConjured(item)) {
      decreaseQuality(item, DOUBLE_QUALITY_UNIT);
    } else {
      decreaseQuality(item);
    }
  }

  /**
   * @param item
   * @return
   */
  private boolean isConjured(Item item) {

    return CONJURED_ITEM.equals(item.getName());
  }

  /**
   * @param item
   * @return
   */
  private boolean isAgedBrie(Item item) {

    boolean isAgedBrie = AGED_BRIE.equals(item.getName());
    return isAgedBrie;
  }

  /**
   * @param item
   * @return
   */
  private boolean isPasses(Item item) {

    boolean isPasses = BACKSTAGE_PASSES.equals(item.getName());
    return isPasses;
  }

  private void setZeroQuality(Item item) {

    item.setQuality(0);
  }

  private boolean hasExpired(Item item) {

    boolean hasExpired = item.getSellIn() < DEADLINE_SELL_IN;
    return hasExpired;
  }

  private void decreaseSellIn(Item item) {

    if (isSulfuras(item))
      return;
    item.setSellIn(item.getSellIn() - SELL_IN_UNIT);
  }

  /**
   * @param name
   * @return
   */
  private boolean isSulfuras(Item item) {

    return SULFURAS.equals(item.getName());
  }

  private boolean notMaxQuality(Item item) {

    boolean notMaxQuality = item.getQuality() < MAX_QUALITY;
    return notMaxQuality;
  }

  private boolean isMinQuality(Item item) {

    boolean isMinQuality = item.getQuality() <= MIN_QUALITY;
    return isMinQuality;
  }

  private void increaseQuality(Item item) {

    boolean isMaxQuality = !notMaxQuality(item);
    if (isMaxQuality)
      return;
    item.setQuality(item.getQuality() + QUALITY_UNIT);
  }

  private void decreaseQuality(Item item, int decreaseUnits) {

    item.setQuality(item.getQuality() - decreaseUnits);
  }

  private void decreaseQuality(Item item) {

    if (isSulfuras(item))
      return;

    if (isMinQuality(item))
      return;

    decreaseQuality(item, QUALITY_UNIT);
  }
}

File: 03-GildedRose/python/fbougeno/gilded_rose.py
# -*- coding: utf-8 -*-
class GildedRose(object):
    """Documentation for GildedRose

    """

    def __init__(self, items):
        super(GildedRose, self).__init__()
        self.items = items
        self.BACKSTAGE = "Backstage passes to a TAFKAL80ETC concert"
        self.SULFURAS = "Sulfuras, Hand of Ragnaros"
        self.AGED_BRIE = "Aged Brie"
        self.CONJURED = "Conjured"
        self.MAX_ITEM_QUALITY = 50
        self.MIN_ITEM_QUALITY = 0
        self.MIN_SELL_IN = 0
        self.REGULAR_QUALITY_DELTA = 1
        self.DOUBLE_QUALITY_DELTA = 2
        self.TRIPLE_QUALITY_DELTA = 3

    def update_quality(self):
        for item in self.items:
            self.perform_daily_quality_upgrade(item)
            self.decrease_sell_in(item)
            self.compute_extra_quality_based_on_sell_in(item)

    def decrease_quality(self, item, delta):
        if self.is_sulfuras(item):
            return
        new_quality = item.quality - delta
        if new_quality > self.MIN_ITEM_QUALITY:
            item.quality = new_quality
        else:
            item.quality = self.MIN_ITEM_QUALITY

    def decrease_sell_in(self, item):
        if self.is_sulfuras(item):
            return
        item.sell_in -= 1

    def increase_quality(self, item, delta):
        if self.is_sulfuras(item):
            return
        new_quality = item.quality + delta
        if new_quality < self.MAX_ITEM_QUALITY:
            item.quality = new_quality
        else:
            item.quality = self.MAX_ITEM_QUALITY

    def increase_special_items_quality(self, item):
        delta_to_apply = self.REGULAR_QUALITY_DELTA
        if item.name == self.BACKSTAGE:
            delta_to_apply = self.get_quality_delta_for_backstage(item)
        self.increase_quality(item, delta_to_apply)

    def get_quality_delta_for_backstage(self, item):
        if item.sell_in < 6:
            return self.TRIPLE_QUALITY_DELTA
        if item.sell_in < 11:
            return self.DOUBLE_QUALITY_DELTA
        return self.REGULAR_QUALITY_DELTA

    def perform_daily_quality_upgrade(self, item):
        if self.is_special_item(item):
            self.increase_special_items_quality(item)
            return

        self.decrease_item_quality_based_on_name(item)

    def decrease_item_quality_based_on_name(self, item):
        quality_delta = self.REGULAR_QUALITY_DELTA
        if self.is_conjured(item):
            quality_delta = quality_delta * 2
        self.decrease_quality(item, quality_delta)

    def compute_extra_quality_based_on_sell_in(self, item):
        if item.sell_in > self.MIN_SELL_IN:
            return

        if self.is_aged_brie(item):
            self.increase_quality(item, self.REGULAR_QUALITY_DELTA)
            return

        if self.is_backstage(item):
            item.quality = item.quality - item.quality
            return

        self.decrease_item_quality_based_on_name(item)

    def is_special_item(self, item):
        return self.is_aged_brie(item) or self.is_backstage(item)

    def is_aged_brie(self, item):
        return item.name == self.AGED_BRIE

    def is_backstage(self, item):
        return item.name == self.BACKSTAGE

    def is_sulfuras(self, item):
        return item.name == self.SULFURAS

    def is_conjured(self, item):
        return self.CONJURED in item.name


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
